optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc

```diff
@@ -58,55 +45,283 @@ that uses a functional approach:
-and supports <<explainingTheScore,justifications>>:
 ////
+TODO Include and describe the following image when PLANNER-1709 is fixed.
+image::ConstraintStreams/constraintStreamJustification.png[align="center"]
+////
+
+
+[[constraintStreamsFirst]]
+== Your First Constraint Stream
+
+To use the ConstraintStreams API in your project, first write a pure Java `ConstraintProvider` implementation like this:
 
 [source,java,options="nowrap"]
 ----
-    private Constraint speakerUnavailableTimeslot(ConstraintFactory factory) {
-        return factory.from(Talk.class)
-                .filter(Talk::hasAnyUnavailableSpeaker)
-                .penalize("Speaker unavailable timeslot",
-                        HardSoftScore.ONE_SOFT,
-                        Talk::getDurationInMinutes);
+    public class MyConstraintProvider implements ConstraintProvider {
+
+        @Override
+        public Constraint[] defineConstraints(ConstraintFactory factory) {
+            return new Constraint[] { // Provide as many constraints as you like. This example only provides one.
+                    firstConstraint(factory)
+            };
+        }
+
+        private Constraint firstConstraint(ConstraintFactory factory) {
+            return factory.from(...) // Choose the initial source of the constraint stream.
+                // Add as many building blocks as you need.
+                .penalize(...); // And finally define the penalty associated with this constraint.
+        }
+
+    }
+----
+
+Then use the following in your solver config:
+
+[source,xml,options="nowrap"]
+----
+    <solver>
+      <scoreDirectorFactory>
+        <constraintProviderClass>com.example.MyConstraintProvider</constraintProviderClass>
+      </scoreDirectorFactory>
+      ...
+    </solver>
+----
+
+
+[[constraintStreamsCardinality]]
+== Constraint Stream Cardinality
+
+Stream cardinality is a measure of how many objects a single constraint match consists of.
+The simplest constraint stream has a cardinality of 1, meaning each constraint match only consists of 1 object.
+Therefore, it is called a `UniStream`:
+
+[source,java,options="nowrap"]
+----
+    private Constraint doNotAssignAnn(ConstraintFactory factory) {
+        return factory.from(Shift.class) // Returns UniStream<Shift>.
+                ...
+    }
+----
+
+Some constraint stream <<constraintStreamsBuildingBlocks, building blocks>> can increase stream cardinality, such as
+<<constraintStreamsJoin,join>> or <<constraintStreamsGroupingAndCollectors,groupBy>>:
+
+[source,java,options="nowrap"]
+----
+    private Constraint doNotAssignAnn(ConstraintFactory factory) {
+        return factory.from(Shift.class) // Returns Uni<Shift>.
+                .join(Employee.class)    // Returns Bi<Shift, Employee>.
+                .join(DayOff.class)      // Returns Tri<Shift, Employee, DayOff>.
+                .join(Country.class)     // Returns Quad<Shift, Employee, DayOff, Country>.
```

Comment:
FYI All references to quad-streams need to disappear if we intend this doc for product 7.6. That feature is not present there.

Reply:
Agreed, only applies when we backport these docs changes to the 7.30.x branch. Do we intend to do that?